### DISCUSSION TITLE
How to avoid truncating leading 0 from cell value from a .csv file

### DISCUSSION THREAD

**Discussion by Tri-Vi** [exceljs/exceljs#2489] [Q&A]
**Problem:**
I use this library to read a .csv file. Inside my .csv file, there is a cell with a leading zero value (0001234). When .csv file is read, the leading 0 value cell gets truncated and becomes a 1234.

**Question**:
Is there a way to avoid that truncate leading 0 issue when read .csv file ?

**Comment by boyepanthera** [comment]
I ran into similar issue, I think the problem is from fast-csv the csv reader used by ExcelJS under the hood, what I did was write my own [custom npm package](https://www.npmjs.com/package/line-by-line-reader) to read the csv line by line - then write a function to split each line into an array based on your desired indexes.


The arrays get pushed as values into worksheet created with Exceljs, Below is code to explain the steps above

```Js
import Excel from 'exceljs';
import lineReader from 'line-by-line-reader';

//Function to split content at indexes
const splitAtIndexes = (string, indexes) => {
    const output = [];
    output.push(string.substring(0, indexes[0]));
    indexes.map((index, i) => {
        if (i > 0) {
            output.push(string.substring(indexes[i - 1], index));
        }
    });
    return output;
};

const workBook = new Excel.Workbook();
let newWorkSheet = workBook.addWorksheet('Worksheet Name');
const filepath ='path/to/your/file'

//Read your file with line by line reader package
lineReader(filepath)
    .then(resultfile=>{
        //To remove the empty first line kept in the array by default in lineReader package
        resultfile.lines.shift();
        resultfile.lines.forEach(line => {
            //split each line of your file based on the index of your data
            let splittedRow = splitAtIndexes(line, [12, 24, 66, 88, 90, 114, 117, 142, 146])
            newWorkSheet.addRow(splittedRow);
        }
})

```

**Comment by mohsinamjad** [comment]
+1

**Comment by willyou** [comment]
+1

**Comment by Airdevelopments** [comment]
As we have faced the same issue, I could actually trace it down to exceljs:

https://github.com/exceljs/exceljs/blob/5bed18b45e824f409b08456b59b87430ded023ab/lib/csv/csv.js#L61

In the given function every cell value is attempted to be parsed as a Number. Therefore a string value like `'000123'` will successfully be parsed to the number `123`.

We resolved the issue temporarily using patch-package to apply a patch that simply removes the attempt to parse a given string and just return it as-is.

It would be nice to be able to control this parsing behavior / provide a custom parsing function to exceljs in the future.
